synapse/handlers/pagination.py

```diff
@@ -176,6 +136,82 @@ def __init__(self, hs: "HomeServer"):
+        """This takes care of restoring unfinished purge/shutdown rooms from the DB.
+        It also takes care to launch scheduled ones, like rooms that has been fully
+        forgotten.
+
+        It should be run regularly.
+        """
+        rooms_to_purge = await self.store.get_rooms_to_purge()
+        for r in rooms_to_purge:
+            room_id = r["room_id"]
+            delete_id = r["delete_id"]
+            status = r["status"]
+            timestamp = r["timestamp"]
+
+            if (
+                status == DeleteStatus.STATUS_COMPLETE
+                or status == DeleteStatus.STATUS_FAILED
+            ):
+                # remove the delete from the list 24 hours after it completes or fails
+                time_since_completed = self.clock.time_msec() - timestamp
+                if time_since_completed >= PaginationHandler.CLEAR_PURGE_AFTER_MS:
+                    await self.store.delete_room_to_purge(room_id, delete_id)
+
+                    del self._delete_by_id[delete_id]
+                    self._delete_by_room[room_id].remove(delete_id)
+                    if not self._delete_by_room[room_id]:
+                        del self._delete_by_room[room_id]
+
+                continue
+
+            delete_status = self._delete_by_id.get(delete_id)
+            if delete_status is not None:
+                # a delete background task is already running (or has run)
+                # for this delete id, let's ignore it
+                continue
+
+            self._delete_by_id[delete_id] = DeleteStatus()
+            self._delete_by_id[delete_id].status = status
+            self._delete_by_room.setdefault(room_id, []).append(delete_id)
+
+            # restore a shutdown from the DB
+            # also take care of purging if needed
+            shutdown_response = None
+            if status == DeleteStatus.STATUS_SHUTTING_DOWN:
+                shutdown_params = json.loads(r["shutdown_params"])
+                if r["shutdown_response"]:
+                    shutdown_response = json.loads(r["shutdown_response"])
+                await self._shutdown_and_purge_room(
+                    room_id,
+                    delete_id,
+                    shutdown_params=shutdown_params,
+                    shutdown_response=shutdown_response,
+                )
+                continue
+
+            # launch a purge from the DB
+            # it may be an interrupted purge or a scheduled one
+            purge_now = True if status == DeleteStatus.STATUS_PURGING else False
+            if status == DeleteStatus.STATUS_WAIT_PURGE:
+                if timestamp is None or self.clock.time_msec() >= timestamp:
+                    purge_now = True
+
+            # TODO 2 stages purge, keep memberships for a while so we don't "break" sync
+            if purge_now:
+                await self.purge_room(
+                    room_id, delete_id, True, shutdown_response=shutdown_response
```

Comment:
This is perhaps dangerous. I am not sure if/what we want to check before re-scheduling a forced purge.
If the room has been joined by an user in the meantime we may want to skip it ?

Same concerns applies to the call to `_shutdown_and_purge_room` since it can also take `force_purge`.

Reply:
@MadLittleMods what do you think of this ?

Reply:
> If the room has been joined by an user in the meantime we may want to skip it ?

Seems like we should do this. Whenever we see a room join, also check this table and `delete_room_to_delete(...)`

Maybe we also want to get a [write lock](https://github.com/matrix-org/synapse/pull/15782) before starting the delete.